storage: Guard against replica removal in splitTrigger's async campaign

d4l3k opened the pull request:
This adds a check after splitting a range that the replica wasn't transferred/replaced in the short time between it being added and the async campaigning.

See #4475

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6634)

<!-- Reviewable:end -->

d4l3k committed:
Added guard to starting a new raft campaign after splitting a range in case it was immediately transferred/replaced and associated tests.

d4l3k commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[storage/client_raft_test.go, line 1025 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHVV4N2hl_HfJDGqtd7:-KHVV4N2hl_HfJDGqtd8:391502538) ([raw file](https://github.com/cockroachdb/cockroach/blob/88a78f7345406f132e1f1aaecbf216e04beea3e1/storage/client_raft_test.go#L1025)):*
> ```Go
> 
> 		states := make(map[raft.StateType]bool)
> 		for ti := time.Now(); time.Now().Sub(ti) < 100*time.Millisecond; {
> ```

Is this checking too aggressive/high CPU usage?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634)*
<!-- Sent from Reviewable.io -->

BramGruneir commented on storage/replica_command.go:
```diff
@@ -2194,7 +2194,21 @@ func (r *Replica) splitTrigger(
+				foundReplica := false
+				if err == nil {
+					for _, rep := range split.NewDesc.Replicas {
+						if rep.ReplicaID == replica.mu.replicaID {
```
doesn't calling anything under `replica.mu` require a lock?

d4l3k replied:
Yeah. Added locks to it and switched `newRng` to be `replica` because we've established that they're the same at the end of the function.

BramGruneir commented on storage/client_raft_test.go:
```diff
@@ -970,6 +970,76 @@ func TestReplicateAfterSplit(t *testing.T) {
+
+		states := make(map[raft.StateType]bool)
+		for ti := time.Now(); time.Now().Sub(ti) < 100*time.Millisecond; {
+			time.Sleep(1 * time.Millisecond)
```
can this be done with a select instead?

d4l3k replied:
Done.

BramGruneir commented on storage/replica_command.go:
```diff
@@ -2194,7 +2194,21 @@ func (r *Replica) splitTrigger(
+					log.Infof("new replica %d removed before campaigning", r.mu.replicaID)
+					return
+				}
 				newRng.mu.Lock()
```
So, it's still possible for it to be removed between the check for foundReplica and this step.

d4l3k replied:
Should be fixed since replica.mu is locked for the whole function now.

BramGruneir commented:
When you title a commit, please add the section you're working on.
in this case, make the title:
`storage: Guard against replica removal in splitTrigger's async campaign`

tamird commented:
is this change logically one commit?

<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> When you title a commit, please add the section you're working on.
> 
> in this case, make the title:
> 
> `storage: Guard against replica removal in splitTrigger's async campaign`</details>


---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[storage/replica_command.go, line 2210 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHVrqQtAc-JlZt3dpIr:-KHVrqQur0A8b_26G3tb:9781351) ([raw file](https://github.com/cockroachdb/cockroach/blob/425e750eb3d77df3409eaa88f421b4d851538f20/storage/replica_command.go#L2210)):*
> ```Go
> 					}
> 				}
> 				if err != nil || !foundReplica {
> ```

there is no need for this error check, right? it's not possible for foundReplica to be true if err != nil

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-:-KHVrOOwsExzESt0a2BZ:1558484805)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/client_raft_test.go, line 1026 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHVV0gEXntM82wmasKk-r3-1026:-KHWIiyii0dkUqBt4OHh:766851969) ([raw file](https://github.com/cockroachdb/cockroach/blob/d82c9e864f84b6042115ea518335dd9ef4c3970f/storage/client_raft_test.go#L1026)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Done.</details>
is there any reason not to use `util.SucceedsSoon` here?

---

*[storage/client_raft_test.go, line 1018 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHWINZ4bLAyo-MdGxeA:-KHWINZ4bLAyo-MdGxeB:-817995455) ([raw file](https://github.com/cockroachdb/cockroach/blob/425e750eb3d77df3409eaa88f421b4d851538f20/storage/client_raft_test.go#L1018)):*
> ```Go
> 			}
> 		}
> 		if td.replace {
> ```

this should be `else if`, right? you should also document in the struct definition that only one of these may be true.

---

*[storage/client_raft_test.go, line 1024 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHWIaMjtbhkbMg18c-D:-KHWIaMjtbhkbMg18c-E:-366351615) ([raw file](https://github.com/cockroachdb/cockroach/blob/425e750eb3d77df3409eaa88f421b4d851538f20/storage/client_raft_test.go#L1024)):*
> ```Go
> 		}
> 
> 		states := make(map[raft.StateType]bool)
> ```

use `struct{}` as the value type.

---

*[storage/client_raft_test.go, line 1048 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHWIvROfMQLJhLvovQA:-KHWIvROfMQLJhLvovQB:1867749879) ([raw file](https://github.com/cockroachdb/cockroach/blob/425e750eb3d77df3409eaa88f421b4d851538f20/storage/client_raft_test.go#L1048)):*
> ```Go
> 			t.Errorf("%d. unexpected replica state %d", i, state)
> 		}
> 		mtc.Stop()
> ```

you should `defer` this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Sort of. They were things that came up from the CR so I committed them together.

<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is this change logically one commit?</details>


---

Review status: 0 of 2 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[storage/client_raft_test.go, line 1026 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHVV0gEXntM82wmasKk-r3-1026:-KHWLS_VJbf2JLh9YeAn:-1950709084) ([raw file](https://github.com/cockroachdb/cockroach/blob/d82c9e864f84b6042115ea518335dd9ef4c3970f/storage/client_raft_test.go#L1026)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is there any reason not to use `util.SucceedsSoon` here?</details>
Yeah. The async campaign call is specified with a fixed 10ms sleep. You need to wait to verify that it doesn't switch into the campaigning state after the initial check.

---

*[storage/client_raft_test.go, line 1018 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHWINZ4bLAyo-MdGxeA:-KHWLpqmV9ECLWVO4WP_:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/425e750eb3d77df3409eaa88f421b4d851538f20/storage/client_raft_test.go#L1018)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this should be `else if`, right? you should also document in the struct definition that only one of these may be true.</details>
Done.

---

*[storage/client_raft_test.go, line 1024 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHWIaMjtbhkbMg18c-D:-KHWLCg4h14VS9JOVsea:1443748349) ([raw file](https://github.com/cockroachdb/cockroach/blob/425e750eb3d77df3409eaa88f421b4d851538f20/storage/client_raft_test.go#L1024)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> use `struct{}` as the value type.</details>
Done. I just thought using bool was cleaner since you could just do `if !states[state]` instead of `if _, ok := states[state]; !ok`.

---

*[storage/client_raft_test.go, line 1048 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHWIvROfMQLJhLvovQA:-KHWM8WWYBqIW7sxJD5B:946419952) ([raw file](https://github.com/cockroachdb/cockroach/blob/425e750eb3d77df3409eaa88f421b4d851538f20/storage/client_raft_test.go#L1048)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> you should `defer` this.</details>
Defer runs after the function exits and it previously threw errors when I created a new multiTestContext without stopping the previous one.

---

*[storage/replica_command.go, line 2210 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHVrqQtAc-JlZt3dpIr:-KHWMjtT1GT5eBsmnxSd:-14623860) ([raw file](https://github.com/cockroachdb/cockroach/blob/425e750eb3d77df3409eaa88f421b4d851538f20/storage/replica_command.go#L2210)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> there is no need for this error check, right? it's not possible for foundReplica to be true if err != nil</details>
Seemed more explicit to keep err != nil, but I changed it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-:-KHWKBEKksd7XzWX2r0D:1618033533)*
<!-- Sent from Reviewable.io -->

tamird commented:
Consider squashing these commits.

<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Sort of. They were things that came up from the CR so I committed them together.</details>


---

Reviewed 1 of 1 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/client_raft_test.go, line 1026 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHVV0gEXntM82wmasKk-r3-1026:-KHWQ-8UBmnNSriFSt2P:352946349) ([raw file](https://github.com/cockroachdb/cockroach/blob/d82c9e864f84b6042115ea518335dd9ef4c3970f/storage/client_raft_test.go#L1026)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Yeah. The async campaign call is specified with a fixed 10ms sleep. You need to wait to verify that it doesn't switch into the campaigning state after the initial check.</details>
Hm, still seems like you could use util/retry instead of doing things in this more manual way?

---

*[storage/client_raft_test.go, line 1048 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHWIvROfMQLJhLvovQA:-KHWQERPEtmPgfrR-A5h:-63824365) ([raw file](https://github.com/cockroachdb/cockroach/blob/425e750eb3d77df3409eaa88f421b4d851538f20/storage/client_raft_test.go#L1048)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Defer runs after the function exits and it previously threw errors when I created a new multiTestContext without stopping the previous one.</details>
oh, I didn't realize this was in the loop - why can't you reuse the mtc?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-:-KHWRKQR_9mmc6kLf8Nk:295366158)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[storage/client_raft_test.go, line 1026 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHVV0gEXntM82wmasKk-r3-1026:-KHWSc-DWU_nwYcnra9Y:-750350216) ([raw file](https://github.com/cockroachdb/cockroach/blob/d82c9e864f84b6042115ea518335dd9ef4c3970f/storage/client_raft_test.go#L1026)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Hm, still seems like you could use util/retry instead of doing things in this more manual way?</details>
This wants to run at regular intervals; util/retry does exponential backoff. I suppose you could use util/retry with a multiplier of 1, but that would be unusual.

---

*[storage/client_raft_test.go, line 1048 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHWIvROfMQLJhLvovQA:-KHWOTMDNbaXyWqAuzUg:200472195) ([raw file](https://github.com/cockroachdb/cockroach/blob/425e750eb3d77df3409eaa88f421b4d851538f20/storage/client_raft_test.go#L1048)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> oh, I didn't realize this was in the loop - why can't you reuse the mtc?</details>
We often use a helper function for things like this (could be a closure defined inline). This lets you defer the `mtc.Stop()`.

See also #6640: it shouldn't be a problem to have two multiTestContext objects running at once.

---

*[storage/client_raft_test.go, line 992 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHWS53ar_OclelsjbO7:-KHWS53bhD7H9WMe_H0_:-410906538) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b87450fd7c581804eb76c4eec0d708b51f614d/storage/client_raft_test.go#L992)):*
> ```Go
> 		},
> 		{ // Default behavior
> 			expected: []raft.StateType{raft.StateFollower, raft.StateCandidate, raft.StateLeader},
> ```

There's no guarantee that you'll catch the replica in the candidate state. I think we just want to check that the final state is `StateLeader`.

We talked yesterday about checking the Term for the remove/replace cases. Did that not work? 

---

*[storage/client_raft_test.go, line 1021 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHWRF1VweMXzUyv5oQ7:-KHWRF1VweMXzUyv5oQ8:160905101) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b87450fd7c581804eb76c4eec0d708b51f614d/storage/client_raft_test.go#L1021)):*
> ```Go
> 			// Simulate replacement by changing replica ID.
> 			desc := replica2.GetReplica()
> 			desc.ReplicaID += 1234
> ```

I think it would be better to export `Replica.SetReplicaID` and use it here instead of modifying the member variable directly. We could put an exported wrapper in `storage/helpers_test.go` to avoid exporting the method in the main codebase. (exported methods in `*_test.go` can only be seen by tests in the `storage` directory, but they can be seen whether the test is in `package storage` or `package storage_test`)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: 1 of 2 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[storage/client_raft_test.go, line 1026 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHVV0gEXntM82wmasKk-r3-1026:-KHWc3Xikx6_RNuwEthN:-619659563) ([raw file](https://github.com/cockroachdb/cockroach/blob/d82c9e864f84b6042115ea518335dd9ef4c3970f/storage/client_raft_test.go#L1026)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This wants to run at regular intervals; util/retry does exponential backoff. I suppose you could use util/retry with a multiplier of 1, but that would be unusual.</details>
Done. Cleaned this up by checking Term instead of state so we don't have to worry about missing it.

---

*[storage/client_raft_test.go, line 1048 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHWIvROfMQLJhLvovQA:-KHWcF1rgO4hCNt4UCfA:1930462511) ([raw file](https://github.com/cockroachdb/cockroach/blob/425e750eb3d77df3409eaa88f421b4d851538f20/storage/client_raft_test.go#L1048)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> We often use a helper function for things like this (could be a closure defined inline). This lets you defer the `mtc.Stop()`.
> 
> See also #6640: it shouldn't be a problem to have two multiTestContext objects running at once.</details>
Added an enclosure so we can use defer.

---

*[storage/client_raft_test.go, line 992 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHWS53ar_OclelsjbO7:-KHWcbDPG-d4EX4vd2bP:-1469131503) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b87450fd7c581804eb76c4eec0d708b51f614d/storage/client_raft_test.go#L992)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> There's no guarantee that you'll catch the replica in the candidate state. I think we just want to check that the final state is `StateLeader`.
> 
> We talked yesterday about checking the Term for the remove/replace cases. Did that not work? </details>
It just slipped my mind. Refactored to use the Term.

---

*[storage/client_raft_test.go, line 1021 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHWRF1VweMXzUyv5oQ7:-KHWcTo5FAhcJH-BDDI6:518041795) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b87450fd7c581804eb76c4eec0d708b51f614d/storage/client_raft_test.go#L1021)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I think it would be better to export `Replica.SetReplicaID` and use it here instead of modifying the member variable directly. We could put an exported wrapper in `storage/helpers_test.go` to avoid exporting the method in the main codebase. (exported methods in `*_test.go` can only be seen by tests in the `storage` directory, but they can be seen whether the test is in `package storage` or `package storage_test`)</details>
I realize changing the ReplicaID is kind of hacky. I just updated it to actually create a new replica and add it.

---

*[storage/replica_command.go, line 2202 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHVV0gFbgPnoL1tN82W-r3-2202:-KHWcnMClDnUg8AeD8PS:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d82c9e864f84b6042115ea518335dd9ef4c3970f/storage/replica_command.go#L2202)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Yeah. Added locks to it and switched `newRng` to be `replica` because we've established that they're the same at the end of the function.</details>
Done.

---

*[storage/replica_command.go, line 2212 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHVV0gFbgPnoL1tN82W-r3-2212:-KHWcp6Ozg4bY0nONgFP:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d82c9e864f84b6042115ea518335dd9ef4c3970f/storage/replica_command.go#L2212)):*
<details><summary><i>Previously, d4l3k (Tristan Rice) wrote…</i></summary>
> Should be fixed since replica.mu is locked for the whole function now.</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM except for the change to RangeID

<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Consider squashing these commits.</details>


---

Review status: 1 of 2 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[storage/client_raft_test.go, line 1025 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHWnAj-9qhK4b0_NLTE:-KHWnAj0W08v2epH7vvJ:-998431982) ([raw file](https://github.com/cockroachdb/cockroach/blob/181f16b1731f3f2a6de36bd89b8707b3dcff57dc/storage/client_raft_test.go#L1025)):*
> ```Go
> 				// Simulate replacement by adding a new replica.
> 				desc := *replica2.Desc()
> 				desc.RangeID = 0
> ```

Why change the range ID? Doesn't this prevent the new range from being used, so the `replace` and `remove` cases are identical?

---

*[storage/client_raft_test.go, line 1026 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHWnOA1vh5_EOC3YWx_:-KHWnOA1vh5_EOC3YWxa:-143334510) ([raw file](https://github.com/cockroachdb/cockroach/blob/181f16b1731f3f2a6de36bd89b8707b3dcff57dc/storage/client_raft_test.go#L1026)):*
> ```Go
> 				desc := *replica2.Desc()
> 				desc.RangeID = 0
> 				replica3, err := storage.NewReplica(&desc, store0, 0)
> ```

Replica IDs always increase; the new replica ID here should be `desc.NextReplicaID`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-:-KHWnjEN10Mh_2IuxDZp:-1277055491)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[storage/client_raft_test.go, line 1025 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHWnAj-9qhK4b0_NLTE:-KHWx1m_yuBWrkN4qMP8:1493229579) ([raw file](https://github.com/cockroachdb/cockroach/blob/181f16b1731f3f2a6de36bd89b8707b3dcff57dc/storage/client_raft_test.go#L1025)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why change the range ID? Doesn't this prevent the new range from being used, so the `replace` and `remove` cases are identical?</details>
Done. Removed case after discussion.

---

*[storage/client_raft_test.go, line 1026 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHWnOA1vh5_EOC3YWx_:-KHWx0_JA7-8VQrf9s_F:1493229579) ([raw file](https://github.com/cockroachdb/cockroach/blob/181f16b1731f3f2a6de36bd89b8707b3dcff57dc/storage/client_raft_test.go#L1026)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Replica IDs always increase; the new replica ID here should be `desc.NextReplicaID`</details>
Done. Removed case after discussion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> LGTM except for the change to RangeID</details>


---

Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-:-KHX1koaGFINi1xzh-48:1417088721)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[storage/client_raft_test.go, line 980 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHXSK_LkgaCvxWCXJn3:-KHXSK_LkgaCvxWCXJn4:-1647551800) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b87450fd7c581804eb76c4eec0d708b51f614d/storage/client_raft_test.go#L980)):*
> ```Go
> 	testData := []struct {
> 		expected []raft.StateType
> 		// only one of these may be true
> ```

is this no longer true?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
That statement is no longer true. Since one of them was removed in the latest revision

tamird commented:
Can `advance` and `remove` both be `true`?

On Wed, May 11, 2016 at 9:03 PM, Tristan Rice <notifications@github.com>
wrote:

> That statement is no longer true. Since one of them was removed in the
> latest revision
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/6634#issuecomment-218633157>
>

d4l3k commented:
They shouldn't both be true. If the replica is removed the raft term shouldn't advance due to a campaign.

tamird commented:
Right, so then the old comment about only one being true should be restored?

d4l3k commented:
Uh, not necessarily. That was because they did two different things that were mutually exclusive. In this case `advance` is dependent expected change to remove.

bdarnell commented:
`advance` is the expected outcome; `remove` and `replace` were the two behaviors being tested. It might be clearer to rename `advance` to `expectAdvance` and move it to the end of the struct.

d4l3k commented:
Review status: 1 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[storage/client_raft_test.go, line 980 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634#-KHXSK_LkgaCvxWCXJn3:-KH_n3zc4yv-Nv-BncHG:-948779616) ([raw file](https://github.com/cockroachdb/cockroach/blob/51b87450fd7c581804eb76c4eec0d708b51f614d/storage/client_raft_test.go#L980)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is this no longer true?</details>
Renamed advance to expectAdvance as suggested by Ben.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6634)*
<!-- Sent from Reviewable.io -->

d4l3k commented:
@tamird Am I clear to merge?

tamird commented:
LGTM thought that sleep in the test is unfortunate

d4l3k commented:
Yeah, unfortunately it's in the code as well.
